### PR TITLE
Read PDFs from azure container

### PR DESCRIPTION
**Pull request by KarolJastrzebski**
https://tools.hmcts.net/jira/browse/RPE-553

This is an MVP implementation where things are to be added, like parsing zipped files that is done in #16 .

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/resources/application.yaml [hunk 1]
```diff
@@ -44,3 +44,6 @@ accesstoken:
   serviceConfig:
     - serviceName: sscs
       validity: ${SAS_TOKEN_VALIDITY:300} #In seconds
+
+scan:
+  interval: ${SCAN_INTERVAL:30000}
```
Comment (rkondratowicz) [inline src/main/resources/application.yaml]: worth adding unit of time imho

Comment (nitinprabhu) [inline src/main/resources/application.yaml]: You will need to define `SCAN_INTERVAL` in terraform else code cannot be deployed.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Test
+    public void connects_without_exceptions() throws URISyntaxException, StorageException {
+        CloudBlobContainer abc = cloudBlobClient.getContainerReference("abc");
+        abc.createIfNotExists();
+
+        BlobStorageRead blobStorageRead = new BlobStorageRead(cloudBlobClient, pdfsConsumer);
+        blobStorageRead.readBlobs();
+    }
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java]: can you also add a test that checks whether `readBlobs` does what it should do?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/PdfsConsumer.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * Consumes list of PDFs from an envelope (single zip file).
+ */
+@Component
+public class PdfsConsumer implements Consumer<List<PDF>> {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/PdfsConsumer.java]: Why we need a consumer here ?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        this.cloudBlobClient = cloudBlobClient;
+        this.pdfsConsumer = pdfsConsumer;
+    }
+
+    @Scheduled(fixedDelayString = "${scan.interval}")
+    public void readBlobs() {
+        cloudBlobClient.listContainers().forEach(cloudBlobContainer -> {
+            try {
+                String containerName = cloudBlobContainer.getName();
+                CloudBlobContainer container = cloudBlobClient.getContainerReference(containerName);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: confusing names imho. what's the type of `cloudBlobContainer` variable passed in this lambda?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                readBlobsFromContainer(container);
+            } catch (URISyntaxException e) {
+                LOGGER.warn("Invalid URL", e);
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: may be put TO-DO comments here that it's going to be logged into app insights

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: same for others

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            } catch (StorageException e) {
+                LOGGER.warn("Could not obtain container reference", e);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: are you sure those should be only warnings?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    public BlobStorageRead(CloudBlobClient cloudBlobClient, Consumer<List<PDF>> pdfsConsumer) {
+        this.cloudBlobClient = cloudBlobClient;
+        this.pdfsConsumer = pdfsConsumer;
+    }
+
+    @Scheduled(fixedDelayString = "${scan.interval}")
+    public void readBlobs() {
+        cloudBlobClient.listContainers().forEach(cloudBlobContainer -> {
+            try {
+                String containerName = cloudBlobContainer.getName();
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Please log some info stating which container you are processing.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        cloudBlobClient = account.createCloudBlobClient();
+    }
+
+    @Test
+    public void connects_without_exceptions() throws URISyntaxException, StorageException {
+        CloudBlobContainer abc = cloudBlobClient.getContainerReference("abc");
+        abc.createIfNotExists();
+
+        BlobStorageRead blobStorageRead = new BlobStorageRead(cloudBlobClient, pdfsConsumer);
+        blobStorageRead.readBlobs();
```
Comment (nitinprabhu) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java]: No asserts ? May be put something that no exception occurred.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.stream.StreamSupport;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+@Component
+public class BlobStorageRead {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Can it be renamed to `DocumentProcessor` or something else ?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java (hunk 8 above)
Comment (rkondratowicz) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java]: extra points for javadoc comment ;)

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java (hunk 9 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: extra points for javadoc comment ;)

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+                    return processZipFile(blobInputStream);
+                } catch (StorageException e) {
+                    LOGGER.warn("Could not download data", e);
+                } catch (URISyntaxException e) {
+                    LOGGER.warn("Invalid URL", e);
+                }
+                return null;
+            })
+            .filter(Objects::nonNull)
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: you may want to filter out empty lists as well

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            .filter(cloudBlob -> cloudBlob.getName().toLowerCase().endsWith(".zip"))
+            // TODO: notify supplier about non-zip files
+            .map(cloudBlob -> {
+                try {
+                    BlobInputStream blobInputStream = container.getBlockBlobReference(cloudBlob.getName())
+                        .openInputStream();
+
+                    return processZipFile(blobInputStream);
+                } catch (StorageException e) {
+                    LOGGER.warn("Could not download data", e);
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Same here - why are these only warnings?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                    LOGGER.warn("Invalid URL", e);
+                }
+                return null;
+            })
+            .filter(Objects::nonNull)
+            .forEach(pdfsConsumer);
+    }
+
+    private List<PDF> processZipFile(BlobInputStream blobInputStream) {
+        List<PDF> pdfs = new ArrayList<>();
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Alternatively, you could use Stream.generate() for a more declarative approach. Up to you.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+        try (ZipInputStream zis = new ZipInputStream(blobInputStream)) {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Do we know how big those zips are going to be?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+            } catch (StorageException e) {
+                LOGGER.warn("Could not obtain container reference", e);
+            }
+        });
+    }
+
+    private void readBlobsFromContainer(CloudBlobContainer container) {
+        StreamSupport.stream(container.listBlobs().spliterator(), false)
+            .filter(CloudBlob.class::isInstance)
+            .map(CloudBlob.class::cast)
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: would be worth adding a comment on those two lines imho

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java (hunk 2 above)
Comment (lgonczar) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java]: Are there unit tests for that class?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+import uk.gov.hmcts.reform.bulkscanprocessor.services.BlobStorageRead;
+import uk.gov.hmcts.reform.bulkscanprocessor.services.PDF;
+
+import java.net.URISyntaxException;
+import java.security.InvalidKeyException;
+import java.util.List;
+import java.util.function.Consumer;
+
+@RunWith(MockitoJUnitRunner.class)
+public class BlobStorageReadTest {
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java]: Where will these tests be executed?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java (hunk 6 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Not at all and would be keen to escalate to error level.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java [hunk 16]
```diff
@@ -43,7 +59,35 @@ public void connects_without_exceptions() throws URISyntaxException, StorageExce
         CloudBlobContainer abc = cloudBlobClient.getContainerReference("abc");
         abc.createIfNotExists();
 
+        Throwable throwable = catchThrowable(() -> {
+            BlobStorageRead blobStorageRead = new BlobStorageRead(cloudBlobClient, pdfsConsumer);
+            blobStorageRead.readBlobs();
+        });
+
+        assertThat(throwable).doesNotThrowAnyException();
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java]: You have to put the lambda as the argument, not the result. And shouldn't it be `assertThatCode(...`?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java (hunk 15 above)
Comment (lgonczar) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/BlobStorageReadTest.java]: We still need unit tests, covering all the edge cases

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+            .forEach(pdfsConsumer);
+    }
+
+    private List<PDF> processZipFile(BlobInputStream blobInputStream) {
+        List<PDF> pdfs = new ArrayList<>();
+
+        try (ZipInputStream zis = new ZipInputStream(blobInputStream)) {
+            ZipEntry ze;
+            while ((ze = zis.getNextEntry()) != null) {
+                PDF pdf = new PDF(ze.getName(), ByteStreams.toByteArray(zis));
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: will the zip file contain only PDFs?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java [hunk 18]
```diff
@@ -42,12 +42,15 @@ public void readBlobs() {
         cloudBlobClient.listContainers().forEach(cloudBlobContainer -> {
             try {
                 String containerName = cloudBlobContainer.getName();
+                LOGGER.info("Processing {} container", containerName);
                 CloudBlobContainer container = cloudBlobClient.getContainerReference(containerName);
                 readBlobsFromContainer(container);
             } catch (URISyntaxException e) {
                 LOGGER.warn("Invalid URL", e);
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Why is this exception not logged as error?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java (hunk 17 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: In this ticket we are only reading PDFs.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/DocumentProcessorTest.java [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+        verify(pdfsConsumer, times(1)).accept(pdfListCaptor.capture());
+
+        assertThat(pdfListCaptor.getAllValues()).hasSize(1);
+
+        assertThat(pdfListCaptor.getValue())
+            .hasSize(3)
+            .anyMatch(pdf -> "abc.pdf".equals(pdf.getFilename()))
+            .anyMatch(pdf -> "def.pdf".equals(pdf.getFilename()))
+            .anyMatch(pdf -> "ghi.pdf".equals(pdf.getFilename()));
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/DocumentProcessorTest.java]: you can replace the 4 lines above with `.containsExactlyInAnyOrder(...)`.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java (hunk 12 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: In my opinion, the benefit would be simplicity. Not that important, though.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java (hunk 13 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Do you know how big individual zip files can be? If they're too big, we can't extract the whole file at once in memory.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/DocumentProcessorTest.java [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    public void connects_without_exceptions() throws URISyntaxException, StorageException {
+        CloudBlobContainer abc = cloudBlobClient.getContainerReference("abc");
+        abc.createIfNotExists();
+
+        Throwable throwable = catchThrowable(() -> {
+            DocumentProcessor documentProcessor = new DocumentProcessor(cloudBlobClient, pdfsConsumer);
+            documentProcessor.readBlobs();
+        });
+
+        assertThat(throwable).doesNotThrowAnyException();
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/DocumentProcessorTest.java]: You have to put the lambda as the argument, not the result. And shouldn't it be assertThatCode(...?

Or expect the throwable to be null.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Scheduled(fixedDelayString = "${scan.interval}")
+    public void readBlobs() {
+        cloudBlobClient.listContainers().forEach(cloudBlobContainer -> {
+            try {
+                String containerName = cloudBlobContainer.getName();
+                LOGGER.info("Processing {} container", containerName);
+                CloudBlobContainer container = cloudBlobClient.getContainerReference(containerName);
+                readBlobsFromContainer(container);
+            } catch (URISyntaxException e) {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: Why is this exception not logged as error?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java (hunk 12 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Oh, and if those zips (or their whole decompressed content) turn out to be too large to be kept in memory, current way won't work, whereas the streamed one will allow for processing the PDFs one by one, without having to extract them all at once.

Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Stream.generate() is an infinite stream and requires limiting. I have had a peek at `ZipFile.streams()` which utilises `Spliterators.spliterator()` and `ZipEntryIterator` where both of them require passing exact number of elements.

Of course, there is always `Stream.builder()` option, but it just another way for building the same array of elements. This brings another idea of converting `BlobInputStream` into `Iterator<PDF>`.

I am keen to hear your thoughts on this.

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: We can just terminate the stream when there's nothing left.

Whether it's a stream or an iterator, doesn't matter. What's important is that we don't read the whole content of the zip file into memory if we don't have to. The agreed limit is already quite big (4*75 MB). If someone uploads a file that's too big to fit in memory, the job will keep failing as long as that file is there.

File: infrastructure/main.tf [hunk 22]
```diff
@@ -51,6 +51,7 @@ module "bulk-scan" {
     FLYWAY_USER                   = "${module.bulk-scan-db.user_name}"
     FLYWAY_PASSWORD               = "${module.bulk-scan-db.postgresql_password}"
     //
+    SCAN_INTERVAL                 = "30000"
```
Comment (lgonczar) [inline infrastructure/main.tf]: Currently, this is the delay between the end of the last job execution and the start of the next. It would be best to either rename it to `SCAN_DELAY` (or sth similar) or set `fixedRateString` in `@Scheduled` annotation.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+    private void readBlobsFromContainer(CloudBlobContainer container) {
+        StreamSupport.stream(container.listBlobs().spliterator(), false)
+            .filter(CloudBlob.class::isInstance)
+            .map(CloudBlob.class::cast)
+            .filter(cloudBlob -> cloudBlob.getName().toLowerCase().endsWith(".zip"))
+            // TODO: notify supplier about non-zip files
+            .map(cloudBlob -> {
+                try {
+                    BlobInputStream blobInputStream = container.getBlockBlobReference(cloudBlob.getName())
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: This should go to try-with-resources, as the stream always has to be closed. Sorry for not mentioning earlier - didn't spot it before.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/PdfsConsumer.java (hunk 3 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/PdfsConsumer.java]: I think this class is not used anymore

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java (hunk 12 above)
Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: I like this but it looks like premature optimisation. Let's address this in separate ticket.

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/BlobStorageRead.java]: Not premature optimisation - we're reading potentially large files into memory and we can see how some input can cause this job to fail. I'm ok with addressing it in another pull request, though.

File: infrastructure/main.tf [hunk 24]
```diff
@@ -51,6 +51,7 @@ module "bulk-scan" {
     FLYWAY_USER                   = "${module.bulk-scan-db.user_name}"
     FLYWAY_PASSWORD               = "${module.bulk-scan-db.postgresql_password}"
     //
+    SCAN_DELAY                    = "30000"
```
Comment (nitinprabhu) [inline infrastructure/main.tf]: we can configure this in `variables.tf` ?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/HandleZipTestDummy.java [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+        assertThat(pdfNames).containsExactlyInAnyOrder(fileNames);
+    }
+
+    void jurisdictionNameIs(String name) {
+        assertThat(jurisdiction).isEqualTo(name);
+    }
+
+    void isInvoked() {
+        assertThat(invoked).isTrue();
+    }
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/HandleZipTestDummy.java]: As mentioned before, let's keep assertions in test. All those method are only called in one place.

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/DocumentProcessorTest.java [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+        CloudBlockBlob blockBlobReference = abc.getBlockBlobReference(testZip);
+        blockBlobReference.uploadFromFile(new File("src/integrationTest/resources/" + testZip).getAbsolutePath());
+
+        HandleZipTestDummy dummy = new HandleZipTestDummy();
+
+        DocumentProcessor documentProcessor = new DocumentProcessor(cloudBlobClient, dummy);
+        documentProcessor.readBlobs();
+
+        dummy.isInvoked();
+        dummy.jurisdictionNameIs("abc");
```
[the remark this answers is not in the document]
Comment (lgonczar) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/features/DocumentProcessorTest.java]: I agree. Naturally you would expect `dummy.isInvoked()` to just return a boolean, not that an assertion will be made. The straight assertion equivalents are short and readable.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.function.BiConsumer;
+import java.util.stream.StreamSupport;
+import java.util.zip.ZipFile;
+
+/**
+ * Main component that is run in intervals.
+ *
+ * Looks up all available containers for zip files and extracts pdf files from them.
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: This class is a processor which reads file from blob storage, extracts zip and stores them in document storage.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DocumentProcessor.class);
+
+    private final CloudBlobClient cloudBlobClient;
+    private final BiConsumer<String, ZipFile> handleZip;
+
+    /**
+     * Create new document processor for zip files stored on a blob storage.
+     *
+     * @param cloudBlobClient connection to Azure Blob Storage
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: Do we need these comments ?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 29]
```diff
@@ -0,0 +1,104 @@
...
+     * @param handleZip callback function invoked for every downloaded zip file.
+     *          See {@link ZipFileHandler#accept(String, ZipFile)}.
+     */
+    @Autowired
+    public DocumentProcessor(CloudBlobClient cloudBlobClient, BiConsumer<String, ZipFile> handleZip) {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: Not sure why we need a BiConsumer.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+     */
+    @Scheduled(fixedDelayString = "${scan.delay}")
+    public void readBlobs() {
+        cloudBlobClient.listContainers().forEach(cloudBlobContainer -> {
+            try {
+                String containerName = cloudBlobContainer.getName();
+                LOGGER.info("Processing {} container", containerName);
+                CloudBlobContainer container = cloudBlobClient.getContainerReference(containerName);
+                readBlobsFromContainer(container);
+            } catch (URISyntaxException e) {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: Can we use Exceptions with Improved Type Checking here ?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+
+    private void readBlobsFromContainer(CloudBlobContainer container) {
+        StreamSupport.stream(container.listBlobs().spliterator(), false)
+
+            // checking if it is safe to cast to CloudBlob
+            .filter(CloudBlob.class::isInstance)
+            .map(CloudBlob.class::cast)
+
+            // so that we could get method for retrieving its name
+            .filter(cloudBlob -> cloudBlob.getName().toLowerCase().endsWith(".zip"))
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: we should not filter zip if we get a non zip exception should be thrown and record should be created in db.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+            // TODO: notify supplier about non-zip files
+            .forEach(cloudBlob -> {
+                File tempFile = null;
+                try {
+                    // let's download every zip file using temporary file
+                    tempFile = File.createTempFile(cloudBlob.getContainer().getName(), "");
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: We could use streams instead of storing it to disk and fetching it again back before sending it to doc store. IO intensive operation. Do we really need to do this?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/ZipFileHandler.java [hunk 33]
```diff
@@ -0,0 +1,48 @@
...
+    public void accept(String jurisdiction, ZipFile zipFile) {
+        try {
+            // extract data from manifest file
+            ZipEntry entry = zipFile.getEntry("manifest.json");
+            InputStream inputStream = zipFile.getInputStream(entry);
+            Envelope envelope = EntityParser.parseEnvelopeMetadata(inputStream);
+
+            // check if all files in the zip are present in the manifest
+            zipFile.stream()
+                .filter(file -> !file.getName().toLowerCase().equals("manifest.json"))
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/ZipFileHandler.java]: file will be called `metadata.json`

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ * Looks up all available containers for zip files and extracts pdf files from them.
+ */
+@Component
+public class DocumentProcessor {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DocumentProcessor.class);
+
+    private final CloudBlobClient cloudBlobClient;
+    private final BiConsumer<String, ZipFile> handleZip;
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: Not that we've got more arguments and everything, maybe it would be better to use a problem-specific interface? By looking at the code you don't know what the string parameter represents.

Also, do we know what the handler will do and need?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java (hunk 31 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: but that would mean the presence of a non-zip file will prevent the processing of other files

Comment (nitinprabhu) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: exception should be thrown while processing a non zip file and other files should continue processing. I would expect exception to be thrown while creating ZipFile object ?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java [hunk 35]
```diff
@@ -0,0 +1,104 @@
...
+                LOGGER.error("Invalid URL", e);
+                // TODO: track exception in AppInsights Telemetry
+            } catch (StorageException e) {
+                LOGGER.error("Could not obtain container reference", e);
+                // TODO: track exception in AppInsights Telemetry
+            }
+        });
+    }
+
+    private void readBlobsFromContainer(CloudBlobContainer container) {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: Worth extracting some logic out of this method - it's grown quite big.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/ZipFileHandler.java [hunk 36]
```diff
@@ -0,0 +1,48 @@
...
+public class ZipFileHandler implements BiConsumer<String, ZipFile> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ZipFileHandler.class);
+
+    @Override
+    public void accept(String jurisdiction, ZipFile zipFile) {
+        try {
+            // extract data from manifest file
+            ZipEntry entry = zipFile.getEntry("manifest.json");
+            InputStream inputStream = zipFile.getInputStream(entry);
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/ZipFileHandler.java]: How about `manitestStream` or sth?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/ZipFileHandler.java [hunk 37]
```diff
@@ -0,0 +1,48 @@
...
+            Envelope envelope = EntityParser.parseEnvelopeMetadata(inputStream);
+
+            // check if all files in the zip are present in the manifest
+            zipFile.stream()
+                .filter(file -> !file.getName().toLowerCase().equals("manifest.json"))
+                .forEach(file -> {
+                    if (!envelope.hasReferenceTo(file.getName())) {
+                        throw new NotImplementedException();
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/ZipFileHandler.java]: What's expected here?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java (hunk 31 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/DocumentProcessor.java]: True. Provided it won't throw a RuntimeException (which it shouldn't)

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/ZipFileHandler.java [hunk 38]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * Instance of this class is automatically injected by Spring into
+ * {@link DocumentProcessor#DocumentProcessor(CloudBlobClient, BiConsumer)} as it implements
+ * {@link BiConsumer}.
+ */
+@Component
+public class ZipFileHandler implements BiConsumer<String, ZipFile> {
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/ZipFileHandler.java]: Needs to be tested

**Comment by KarolJastrzebski** [comment]
Closing this one as @nitinprabhu is getting his take with the code.
